src/mpyl/steps/deploy/k8s/kubectl.py

```diff
@@ -0,0 +1,67 @@
+""" This module is called on to push your kubernetes manifest to a(n) (argo) repo during the `mpyl.steps.deploy step."""
+
+import shutil
+from pathlib import Path
+from tempfile import TemporaryDirectory
+
+from .deploy_config import get_namespace
+from ....project import Target
+from ....steps import Input, Output
+from ....utilities.repo import RepoConfig, Repository
+
+
+def push_manifest_to_repo(
+    step_input: Input,
+    manifest_path: Path,
+) -> Output:
+    git_config = step_input.run_properties.config["vcs"]["argoRepository"]
+    if not git_config:
+        raise ValueError("No argocd repository configured")
+    argocd_repo_config = RepoConfig.from_git_config(git_config=git_config)
+    with TemporaryDirectory() as tmp_repo_dir:
+        with Repository.from_clone(
+            config=argocd_repo_config, repo_path=Path(tmp_repo_dir)
+        ) as argo_repo:
+            branch = "main"
+            if argo_repo.local_branch_exists(branch_name=branch):
+                argo_repo.delete_local_branch(
+                    branch_name=branch
+                )  # To enforce latest version of branch
+            branch_exists = argo_repo.remote_branch_exists(branch_name=branch)
+            if branch_exists:
+                argo_repo.checkout_branch(branch_name=branch)
+            else:
+                argo_repo.create_branch(branch_name=branch)
+
+            folder_name = __get_folder_name(step_input=step_input)
+            new_file_path = Path(
+                tmp_repo_dir,
+                folder_name,
+                get_namespace(step_input.run_properties, step_input.project),
+                step_input.project.name,
+                manifest_path.name,
+            )
+            shutil.copytree(
+                src=manifest_path.parent,
+                dst=new_file_path.parent,
+                dirs_exist_ok=True,
+            )
+
+            if argo_repo.has_changes:
+                argo_repo.stage(".")
+                argo_repo.commit(
+                    f"Committing new manifest for project {step_input.project.name} on environment {folder_name}"
```

Comment:
Is it possible to add either the branch name or the PR number to the commit message?
Something like:
* _Deploy project {step_input.project.name} to environment {folder_name} from master branch_
* _Deploy project {step_input.project.name} to environment {folder_name} from PR-123_ 

This would make the repository history easier to parse.

Reply:
Thinking a bit more about this, I think it would be wise to make deployments atomic, i.e. changes to all applications involved in the same PR or tag to be applied to this repository at the same time (either in the same commit, or in multiple commits in a branch) instead of in multiple operations.
Like [Sam mentioned in a different thread](https://github.com/Vandebron/mpyl/pull/295#discussion_r1381565092), this will make rollbacks trivial as it will mean simply reverting a single commit.

Reply:
I'll try to clear up 

***Production*** The manifest repo is a mirror image of your infrastructure. The manifests in this [folder](https://github.com/SamTheisens/mpyl-example-argocd/tree/main/manifests/projects) are mapped nearly 1-to-1 to applications via [this generator](https://github.com/SamTheisens/mpyl-example-argocd/blob/main/argo/monorepo_application.yaml).
They end up in ArgoCD (and the K8S cluster) as follows:
![Screenshot 2023-11-04 at 09 28 00](https://github.com/Vandebron/mpyl/assets/1911436/14673479-3214-4c21-b619-bd7e3d95b537)
The generator point to `HEAD` of `main`. Any revision pushed to main results in a deployment of one or more applications, depending which applications' manifests were altered.
So ArgoCD deployments are per **repo**, not per application.

***Acceptance*** should be as much as possible identical to production. The acceptance and production generators cannot point to the same folder with manifests, because the manifests will differ (slightly). You could use separate a folder for `test`, `acceptance` and `production`. 

***PR*** for each pull request in your _source code_ repo a counterpart is created in the _artifact_ repo. Only the applications that were altered will receive alterations to the manifests. These manifests are deployed into a namespace that matches the pr number.
Here's an [example PR ](https://github.com/SamTheisens/mpyl-example-argocd/pull/3/files#diff-e842e423095b5fe930c3c15ea147f3b91793e47a3d6f717eddf97f0b8e51dd03). The manifests are mapped to application via [this generator](https://github.com/SamTheisens/mpyl-example-argocd/blob/main/argo/pullrequest_generator.yaml)
And this is what its ArgoCD counterpart looks like:
![Screenshot 2023-11-04 at 09 45 03](https://github.com/Vandebron/mpyl/assets/1911436/9ed8719c-b70c-4533-bb99-34eec6c515ff)

The deployment operations per stage could be done as:
 - _PR_ push a revision to a pull request on the artifacts repo. This PR will never be merged to master. They can be closed periodically to clean up old preview environments.
 - _Acceptance_ build artifacts with MPyL in the usual way and push the artifacts to a branch (with PR for more transparency and convenience) on the artifacts repo. The _merge_ of this PR is what effectively deploys the applications.
 - _Production_ process would be identical to acceptance. Only difference is the target folder in the manifests repo. A rollback would be achieved by: create revert PR, review the manifest diffs, merge it.

I've tried out the PR scenario with `mpyl build artifacts push` https://github.com/SamTheisens/mpyl-example-gha/blob/main/.github/workflows/build-mpyl-pipeline.yml#L61 and this seems to work. Acc and prod could be done the same way. Just a different target folder if I'm not mistaken.

WDYT?

Reply:
I actually meant something different with my previous comment, but there's so much great info in your comment that I'll start with that 😄 

> _Production_ The manifest repo is a mirror image of your infrastructure. The manifests in this [folder](https://github.com/SamTheisens/mpyl-example-argocd/tree/main/manifests/projects) are mapped nearly 1-to-1 to applications via [this generator](https://github.com/SamTheisens/mpyl-example-argocd/blob/main/argo/monorepo_application.yaml).
The generator point to HEAD of main. Any revision pushed to main results in a deployment of one or more applications, depending which applications' manifests were altered.
_Acceptance_ should be as much as possible identical to production. The acceptance and production generators cannot point to the same folder with manifests, because the manifests will differ (slightly). You could use separate a folder for test, acceptance and production.

> _Acceptance_ build artifacts with MPyL in the usual way and push the artifacts to a branch (with PR for more transparency and convenience) on the artifacts repo. The merge of this PR is what effectively deploys the applications.
_Production_ process would be identical to acceptance. Only difference is the target folder in the manifests repo. A rollback would be achieved by: create revert PR, review the manifest diffs, merge it.

This is pretty much the setup we've configured so far, down to the usage of the Git generator with a set of folders per environment. Glad to see we're aligned 💪
The only difference is that currently we're committing directly to the `main` branch of the manifests repository (simply because it was faster to build in MPyL), but I 100% agree with you that creating a PR there is a much better option and I want to invest in that before rolling this out company-wide.


> _PR_ for each pull request in your source code repo a counterpart is created in the artifact repo. Only the applications that were altered will receive alterations to the manifests. These manifests are deployed into a namespace that matches the pr number.
_PR_ push a revision to a pull request on the artifacts repo. This PR will never be merged to master. They can be closed periodically to clean up old preview environments.

This is where our setup differs. For simplicity's sake, we're also using a Git directory generator here (like with the "regular" environments), using a folder named `pr-<number>` as the base (instead of `test`, `acceptance` or `production`). When a PR is closed, we have a Github action that removes that directory from the manifests repository. I have to say I really like your approach with the PR generator, and I'll look into that more closely - thanks for the suggestion and example code!

---

Now, to clarify my previous comment: a single code change in our monorepo might span multiple services, and the way MPyL currently works is by committing once per sub-module. So, currently, one single "deployment" might actually consist of multiple commits to the manifests repository. I believe it would be better to group all these changes, either by:
* making them all part of a single commit; or
* making all the commits part of the same branch / pull request

This would IMHO make it much easier to review and rollback deployments.

Reply:
I think I understand wat you are trying to do. I completely agree, deploys should span all services that were touched in the context of a PR.
But that functionality is _already_ present in the latest version of MPyL :-) Just run `mpyl build artifacts push --pr 1234 --path manifests --artifact-type manifests` at the end of your mpyl run. As demonstrated in this example github pipeline https://github.com/SamTheisens/mpyl-example-gha/blob/main/.github/workflows/build-mpyl-pipeline.yml

Reply:
Small note on the automated removal of closed/merged PR and their preview environments in the test cluster: this may be quicker than convenient. In practice I've found it convenient for the PR environment to linger a while, so you can (for example) replay a test scenario in the PR env to compare it to what happens in acceptance.
A periodic cleanup of closed PRs older than x days is conceptually less clean, but perhaps more workable in practice.

Reply:
I think I solved all comments in this thread, please comment on specific pieces of code if I missed something. Easier to track that way!